luthier/src/main/java/com/yahoo/luthier/webservice/data/config/metric/MetricMakerDictionary.java

```diff
@@ -0,0 +1,213 @@
+        if (metricMakers != null) {
+            // should never happen unless multiple loaders are running in race-condition
+            ConcurrentModificationException e = new ConcurrentModificationException();
+            LOG.error("Multiple loaders updating MetricMakerDictionary", e);
+            throw e;
+        }
+        return true;
+    }
+
+    /**
+     * Adds all of the metric makers in the specified collection to the dictionary.
+     *
+     * @param metricMakers collection of metric makers to add
+     * @return <tt>true</tt> if the dictionary changed as a result of the call
+     */
+    public boolean addAll(Collection<MetricMaker> metricMakers) {
+        boolean flag = false;
+        for (MetricMaker metricMaker : metricMakers) {
+            flag = add(metricMaker.toString(), metricMaker) || flag;
+        }
+        return flag;
+    }
+
+    @Override
+    public String toString() {
+        return "MetricMaker Dictionary: " + nameToMetricMaker;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(nameToMetricMaker);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj instanceof MetricMakerDictionary) {
+            MetricMakerDictionary that = (MetricMakerDictionary) obj;
+            return nameToMetricMaker.equals(that.nameToMetricMaker);
+        }
+        return false;
+    }
+
+    /**
+     * Parse parameters for maker's constructor based on parameter's name and type.
+     *
+     * @param paramType type of the parameter in maker's constructor
+     * @param paramName name of the parameter in maker's constructor
+     * @param maker     the maker template used to find parameter's value by name
+     * @return the value of parameter (can be any type)
+     */
+    private Object parseParams(String paramType, String paramName, MetricMakerTemplate maker) {
+        if ("MetricDictionary".equals(paramType)) {
+            return metricDictionary;
+        }
+        if ("DimensionDictionary".equals(paramType)) {
+            return dimensionDictionary;
+        }
+        if ("ZonelessTimeGrain".equals(paramType)) {
+            return DefaultTimeGrain.valueOf(maker.getParams().get(paramName));
+        }
+        if ("ArithmeticPostAggregationFunction".equals(paramType)) {
+            return ArithmeticPostAggregation
+                    .ArithmeticPostAggregationFunction
+                    .valueOf(maker.getParams().get(paramName));
+        }
+        if ("boolean".equals(paramType)) {
+            return Boolean.parseBoolean(maker.getParams().get(paramName));
+        }
+        if ("int".equals(paramType)) {
+            return Integer.parseInt(maker.getParams().get(paramName));
+        }
+        if ("double".equals(paramType)) {
+            return Double.parseDouble(maker.getParams().get(paramName));
+        }
```

Comment:
Similarly, I'm sure you shouldn't have to do string matching here. There is probably a way to generalize this, which you should go try to figure out. 

We cannot have this generalized because then we are restricting the types of parameters you can pass to metric makers by what is hard coded here. It is very reasonable for somebody to want to pass a string to a metric maker, or even a map or list of strings, which you cannot do here. My instinct is for the template to handle resolving the type, but that may not be right.